design-proposals/klip-55-tablespaces.md

```diff
@@ -1,3 +1,210 @@
+
+- The grammar to support tablespace commands
+- The behavior of those commands
+- Backward compatibility
+
+## What is not in scope
+
+- How it's implemented
+
+## Value/Return
+
+- Improved program performance
+- Isolation of functional parameters to a subset of objects
+
+## Public APIs
+
+**Using the default tablespace**
+
+Creating an object works like normal—it's placed into a default tablespace so
+that you never need to use this feature unless you want to.
+
+```sql
+CREATE stream object WITH (
+    ...
+);
+```
+
+The default tablespace is `ksql_default`.
+
+**Creating a tablespace**
+
+You can supply one or more key/value parameters. These parameters are server-level
+configuration parameters that you want to group together.
+
+```sql
+CREATE TABLESPACE tablespace_name WITH (
+    property_name = expression [,
+    ...
+    ]
+);
+```
+
+By default, no objects are placed within a new tablespace.
+
+- TODO: what happens if `tablespace_name` already exists?
+
+**Adding a newly created object to a tablespace**
+
+Explicitly define what tablespace a stream or table should reside in.
+
+```sql
+CREATE [STREAM|TABLE] obj WITH (
+    'tablespace' = 'tablespace_name',
+    ...
+);
+```
+
+If `tablespace_name` doesn't exist, `obj`'s creation is rejected with a clear error message why.
+
+**Seeing what tablespace an object is in**
+
+```sql
+DESCRIBE obj EXTENDED;
+```
+
+Output fragment:
+
+```
+Name                 : KSQL_PROCESSING_LOG
+Type                 : STREAM
+Timestamp field      : Not set - using <ROWTIME>
+Tablespace           : tablespace_name
+```
+
+**Setting a default tablespace**
+
+You can override the default tablespace so you don't
+need to supply it to every proceeding object that gets created.
+
+```sql
+SET 'ksql.default.tablespace' = 'tablespace_name';
+```
+
+If `tablespace_name` doesn't exist, `obj`'s creation is rejected with a clear error message why.
+
+**Seeing the current default tablespace**
+
+```sql
+SHOW PROPERTIES;
+```
+
+TODO: can we get more granular than that?
+
+**Changing an existing object's tablespace**
+
+```sql
+ALTER <STREAM|TABLE> obj_name SET 'tablespace' = 'tablespace_name';
+```
+
+**Changing a tablespace's configuration**
+
+Causes all objects in the tablespace to be immediately reconfigured
+with the new parameters.
+
+```sql
+ALTER TABLESPACE tablespace_name SET 'key' = 'value';
+```
+
+TBD: should this support bulk updating parameters?
+
+**Viewing a tablespace's configuration**
+
+```sql
+DESCRIBE tablespace_name [EXTENDED];
+```
+
+Output:
+
+```sql
++-----------+--------+
+| parameter | value  |
++-----------+--------+
+|       k1  | v1     |
+|       k2  | v2     |
+|       k3  | v3     |
++-----------+--------+
+```
+
+**Dropping a tablespace**
+
+`DROP` only succeeds if it has no objects in it. If it has objects, it lists
+out which ones need to be dropped first.
+
+```sql
+DROP TABLESPACE [IF EXISTS] tablespace_name;
+```
+
+**Seeing all objects in a tablespace**
+
+TBD.
```

Comment:
i would expect the extant `show streams|tables` command to show me only objects from the current space

Reply:
Do you mean that `SHOW <tablespace>` name should do what the header indicates, or something else? +1 on the former, that's a great idea.

Reply:
i don't think `SHOW space-name` is required for anything ? It doesn't fit the pattern of what a `SHOW` or `LIST` generally does, which is to enumerate all objects of a type, rather than describe a specific object or its contents. We don't, for example, intuitively expect `SHOW my-specific-stream-name` to be a thing we can ever do, so why would it work for streamspaces ?

Reply:
i would totally support `SHOW STREAMSPACES` to get a list of all of them though!   (you might have that covered already someplace else in the doc)

Reply:
actually, thinking some more, we should perhaps think twice about my original comment here (that `show streams|tables` should only enumerate objects from the current space). While I personally like the idea, it does edge towards making the space be more of a 'container' than perhaps you were intending to advocate for in this particular KLIP ? We perhaps need a position on what the eventual containment and namespacing model of the system wants to be prior to 'accidentally' starting to back into it. Not sure it does much harm either, just floating that thought.  Different databases approach this somewhat differently so there isn't necessarily one prevalent model to follow. 
(I just happen to personally like the MSSQL flavor, which kind of smushes together the "physical storage properties" and "namespacing boundary" behaviors into a single construct. IMHO this is more intuitive but YMMV. Don't want this bias to leak over into your more surgical proposal here)